### PR TITLE
Core & Internals: added grouped FIFO release and volumn check per destination RSE to throttler & tests; Fix #2220

### LINKED ISSUE #2220: Throttler mechanic to release transfers based on a strategy (FIFO)
Motivation
----------
Right now, the throttler releases transfers simply based on time. FIFO. For the Tape Carousel activity, it is necessary to release them in a smarter way: Grouped FIFO. This essentially means, if a transfer is being released, it should also release all the transfers of the same dataset, so they are submitted to FTS in the same time-window to allow subsequent grouping on FTS side, and thus on the tape system.

### PR DESCRIPTION
**Pull request by hahahannes**
Core & Internals: added grouped FIFO release and volumn check per destination RSE to throttler & tests; Fix #2220

### CODE REVIEW

**Comment by hahahannes** [comment]
I am bit struggling with the sqlite tests. It seems that python uses its own internal version based on version 3.7 but in order to get my tests working I need sqlite 3.25 wich brings window function support. 
Also it seems like there is some work going on that:
https://bugs.python.org/issue34916

**Comment by hahahannes** [comment]
Also this would be the query for selecting the IDs (without the update statement):
``` sql
SELECT id
FROM (
	SELECT 
		ATLAS_RUCIO.REQUESTS.id,
		ATLAS_RUCIO.REQUESTS.requested_at,
                ATLAS_RUCIO.REQUESTS.bytes,
		NVL(t1.name, ATLAS_RUCIO.REQUESTS.name) as dataset_name,
		NVL(t1.SCOPE, ATLAS_RUCIO.REQUESTS.scope) as dataset_scope,
		ATLAS_RUCIO.REQUESTS.name AS name,
		ATLAS_RUCIO.REQUESTS.SCOPE AS scope
	FROM ATLAS_RUCIO.REQUESTS LEFT JOIN (
		SELECT child_name, name, child_scope, scope, row_number() OVER (PARTITION BY child_name, child_scope ORDER BY created_at) as rn
		FROM ATLAS_RUCIO.CONTENTS
	) t1 ON ATLAS_RUCIO.REQUESTS.name = t1.child_name AND ATLAS_RUCIO.REQUESTS.scope = t1.child_scope AND t1.rn = 1
	WHERE ATLAS_RUCIO.REQUESTS.state = 'W' AND ATLAS_RUCIO.REQUESTS.dest_rse_id = HEXTORAW('C34561EEF2D64D75A02F33809E0FC928')
) filtered_requests_subquery
JOIN (
	SELECT
		name,
		scope,
		amount_childs,
		SUM(amount_childs) OVER (ORDER BY oldest_req_at) as cum_amount_childs
	FROM (
		SELECT 
			MIN(requested_at) as oldest_req_at,
			SUM(bytes) as volumn,
			COUNT(*) as amount_childs,
			name,
			scope
		FROM (
			SELECT
				NVL(t1.SCOPE, ATLAS_RUCIO.REQUESTS.scope) AS scope, 
				NVL(t1.name, ATLAS_RUCIO.REQUESTS.name) AS name,
				ATLAS_RUCIO.REQUESTS.bytes,
				ATLAS_RUCIO.REQUESTS.requested_at
			FROM ATLAS_RUCIO.REQUESTS LEFT JOIN (
				SELECT child_name, name, child_scope, scope, row_number() OVER (PARTITION BY child_name, child_scope ORDER BY created_at) as rn
				FROM ATLAS_RUCIO.CONTENTS
			) t1
			ON ATLAS_RUCIO.REQUESTS.name = t1.child_name AND ATLAS_RUCIO.REQUESTS.scope = t1.child_scope AND rn = 1
			WHERE ATLAS_RUCIO.REQUESTS.state = 'W' AND ATLAS_RUCIO.REQUESTS.dest_rse_id = HEXTORAW('C34561EEF2D64D75A02F33809E0FC928')
		)
		GROUP BY SCOPE, name 
	) grouped_requests_subquery
) grouped_cumulated ON filtered_requests_subquery.dataset_name = grouped_cumulated.name AND filtered_requests_subquery.dataset_scope = grouped_cumulated.SCOPE
WHERE grouped_cumulated.cum_amount_childs - grouped_cumulated.amount_childs < [threshold]
```

**Comment by hahahannes** [comment]
> I am bit struggling with the sqlite tests. It seems that python uses its own internal version based on version 3.7 but in order to get my tests working I need sqlite 3.25 wich brings window function support.
> Also it seems like there is some work going on that:
> https://bugs.python.org/issue34916

This is fixed by compiling the correct sqlite version in the dockerfile

File: lib/rucio/core/did.py [hunk 1]
```diff
@@ -464,7 +464,8 @@ def __add_collections_to_container(scope, name, collections, account, session):
         elif match('.*IntegrityError.*ORA-00001: unique constraint .*CONTENTS_PK.*violated.*', error.args[0]) \
                 or match('.*IntegrityError.*1062.*Duplicate entry .*for key.*PRIMARY.*', error.args[0]) \
                 or match('.*columns scope, name, child_scope, child_name are not unique.*', error.args[0]) \
-                or match('.*IntegrityError.*duplicate key value violates unique constraint.*', error.args[0]):
+                or match('.*IntegrityError.*duplicate key value violates unique constraint.*', error.args[0]) \
+                or match('.*IntegrityError.* UNIQUE constraint failed: contents.scope, contents.name, contents.child_scope, contents.child_name.*', error.args[0]):
```
Comment (hahahannes) [inline lib/rucio/core/did.py:468]: It seems that sqlite changed the error message in the latest version, so this was not catched and let to an error in the tests

File: etc/docker/travis/Dockerfile [hunk 2]
```diff
@@ -32,6 +32,18 @@ RUN yum install -y python-pip gmp-devel krb5-devel httpd mod_ssl mod_auth_kerb m
 RUN yum install -y python34 python34-pip python34-devel
 RUN rm -rf /usr/lib/python2.7/site-packages/ipaddress*
 
+# Install sqlite3 version 3.28
+RUN curl https://www.sqlite.org/2019/sqlite-autoconf-3280000.tar.gz > sqlite.tar.gz
```
Comment (bari12) [inline etc/docker/travis/Dockerfile:37]: Just a comment: I am getting a bit concerned about all the different dependency versions which no exists outside of the `pip-requires*` files. Now they are in different dockerfiles, and also in the setup.py of the package when doing someting like `pip install rucio[mysql]`. I think we should have a central way to define all these dependencies and just pull them in from there wherever needed. Not sure though what's the best way to do this

File: lib/rucio/core/request.py [hunk 3]
```diff
@@ -887,43 +894,255 @@ def get_stats_by_activity_dest_state(state, session=None):
 
 
 @transactional_session
-def release_waiting_requests(rse, activity=None, rse_id=None, count=None, account=None, session=None):
+def release_waiting_requests_per_free_volumn(rse, rse_id=None, max_transfer_volumn=None, session=None):
```
Comment (bari12) [inline lib/rucio/core/request.py]: Change all the mentions of `volumn` to `volume`

File: lib/rucio/core/rse.py [hunk 4]
```diff
@@ -657,6 +657,7 @@ def set_rse_transfer_limits(rse, activity, rse_id=None, rse_expression=None, max
     :param max_transfers: Maximum transfers.
     :param transfers: Current number of tranfers.
     :param waitings: Current number of waitings.
+    :param max_volumn: Maximum transfer volumn in bytes.
```
Comment (bari12) [inline lib/rucio/core/rse.py]: volume

File: lib/rucio/daemons/conveyor/throttler.py [hunk 5]
```diff
@@ -149,86 +152,111 @@ def __schedule_requests():
...
-                        set_rse_transfer_limits(rse=None, activity=activity, rse_id=dest_rse_id, max_transfers=threshold, transfers=transfer, waitings=waiting)
-                        record_gauge('daemons.conveyor.throttler.set_rse_transfer_limits.%s.%s.max_transfers' % (activity, rse_name), threshold)
-                        record_gauge('daemons.conveyor.throttler.set_rse_transfer_limits.%s.%s.transfers' % (activity, rse_name), transfer)
-                        record_gauge('daemons.conveyor.throttler.set_rse_transfer_limits.%s.%s.waitings' % (activity, rse_name), waiting)
+                    result_dict[dest_rse_id]['activities'][activity]['accounts'][account]['transfer'] += counter
+                    result_dict[dest_rse_id]['activities'][activity]['transfer'] += counter
+                    result_dict[dest_rse_id]['transfer'] += counter
+
+        for dest_rse_id in result_dict:
+            dest_rse_release_strategy = config_core.get('throttler_release_strategy', 'dest_%s' % dest_rse_id, default='grouped_fifo', use_cache=False)
```
Comment (bari12) [inline lib/rucio/daemons/conveyor/throttler.py]: Maybe the default should be fifo?

File: lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def upgrade():
+    '''
+    Upgrade the database to this revision
+    '''
+
+    if context.get_context().dialect.name in ['oracle', 'mysql', 'postgresql']:
+        schema = context.get_context().version_table_schema if context.get_context().version_table_schema else ''
+        add_column('rse_transfer_limits', sa.Column('max_volumn', sa.BigInteger), schema=schema)
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py]: volume

File: lib/rucio/db/sqla/models.py [hunk 7]
```diff
@@ -663,6 +663,7 @@ class RSETransferLimit(BASE, ModelBase):
     activity = Column(String(50))
     rse_expression = Column(String(3000))
     max_transfers = Column(BigInteger)
+    max_volumn = Column(BigInteger)
```
Comment (bari12) [inline lib/rucio/db/sqla/models.py]: volume
Please also add to etc/sql/.../schema.sql

**Review by bari12 — CHANGES_REQUESTED** [review]

File: lib/rucio/daemons/conveyor/throttler.py (hunk 5 above)
Comment (hahahannes) [reply lib/rucio/daemons/conveyor/throttler.py]: yes, makes sense. Then its easier to update to the next release.

File: lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+# Copyright 2013-2019 CERN for the benefit of the ATLAS collaboration.
```
Comment (bari12) [inline lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py]: One last change request from me (I promise)
Please also change the file name from `volumn` to `volume`.

**Review by bari12 — CHANGES_REQUESTED** [review]

File: lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py (hunk 8 above)
Comment (hahahannes) [reply lib/rucio/db/sqla/migrate_repo/versions/2cbee484dcf9_added_column_max_volumn_to_rse_transfer_.py]: ah yes, forgot that one

**Review by bari12 — APPROVED** [review]

**Pull request merged**